Dart Solution 2

TarishAhmed opened the pull request:
## Description

This PR contains a Dart solutions:

- Implement loop unrolling and memory optimizations
- Add 1-bit and 8-bit variants with parallel versions
- Improve performance over original solution
- Inspired by solution_1's approach

## Contributing requirements

- [x] I read the contribution guidelines in `CONTRIBUTING.md`.
- [x] I placed my solution in the correct solution folder.
- [x] I added a `README.md` with the right badge(s).
- [x] I added a `Dockerfile` that builds and runs my solution.
- [x] I selected `drag-race` as the target branch.
- [x] All code herein is licensed compatible with BSD-3.

rbergen commented on PrimeDart/solution_2/bin/prime_1_bit.dart:
```diff
@@ -0,0 +1,127 @@
+
+class PrimeSieve {
+  final int _sieveSize;
+  final Uint8List _bits;
```
This implementation is using an 8-bit list as the prime candidate container. That makes it an 8-bit implementation, not a 1-bit one. Either modify the implementation to actually use 1 bit per prime, or change the implementation's classification.

rbergen commented on PrimeDart/solution_2/bin/prime_8_bit_par.dart:
```diff
@@ -0,0 +1,46 @@
+
+  while (true) {
+    // Instead of creating a new sieve, just clear the existing one.
+    sieve.clear();
```
This clears the sieve, but doesn't recreate it. As per the contributing guidelines, the sieve needs to be created from scratch (memory allocation included) in every iteration of the main loop for the implementation to be faithful.

rbergen commented on PrimeDart/solution_2/bin/prime_8_bit.dart:
```diff
@@ -0,0 +1,144 @@
+
+    stdout.write(
+        'tarish_8bit;$passes;$duration;1;algorithm=base,faithful=yes,bits=8\n');
+    stdout.writeln('Passes/s/t: ${passes / (duration * 1)}');
```
This line should be sent to stderr.

rbergen commented on PrimeDart/solution_2/bin/prime_8_bit_par.dart:
```diff
@@ -0,0 +1,46 @@
+  stdout.writeln(
+      'tarish_8bit_par;$passes;$time;$processors;algorithm=base,faithful=yes,bits=8');
+
+  stdout.writeln('Passes/s/t: ${passes / (time * processors)}');
```
This line should be sent to stderr.

rbergen commented on PrimeDart/solution_2/bin/prime_1_bit_par.dart:
```diff
@@ -0,0 +1,150 @@
+  stdout.writeln(
+      'tarish_1bit_par;$passes;$time;$processors;algorithm=base,faithful=yes,bits=1');
+  
+  stdout.writeln('Passes/s/t: ${passes / (time * processors)}');
```
This line should be sent to stderr.

rbergen reviewed (changes requested):
Thanks for your submission.

You are effectively submitting this solution as an improvement of PrimeDart/solution_1; the main characteristics of your solution's implementations (at least as advertised) match those of that solution.

As per the contributing guidelines that means the PR should be structured as a proposal to improve the existing solution, instead of adding a new one.

On top of that, there are some concrete issues with the implementations you're submitting. I'm including the comments on those so you can consider them while restructuring your PR along the lines of what I just wrote.

TarishAhmed commented:
@rbergen Thanks for pointing out the issues

I’ve updated the PR to incorporate all the feedback and issues you raised. Specifically:

✅ I’ve addressed the issue of prime_1_bit using 8bit, by modifying the implementation to use 1 bit per 
✅ I have removed the code which cleared the sieve and have sticked to recreating the 
✅ I have changed all stdout to stderr

These changes ensure that the PR now aligns with the expected behavior and coding standards. Please review the latest commit when you have a moment, and let me know if anything else needs further refinement.

Thank you for your guidance.

rbergen commented on PrimeDart/solution_1/bin/prime_8_bit.dart:
```diff
@@ -62,17 +62,17 @@ class PrimeSieve {
+    10: 4, 
+    100: 25, 
+    1000: 168, 
+    10000: 1229, 
     100000: 9592,
-    1000000: 78498,
-    10000000: 664579,
+    1000000: 78498, 
+    10000000: 664579, 
     100000000: 5761455,
-    1000000000: 50847534,
+    1000000000: 50847534, 
```
Whitespace

rbergen commented on PrimeDart/solution_1/bin/prime_8_bit.dart:
```diff
@@ -126,18 +202,16 @@ class PrimeSieve {
-        }
+        // In Dart, using the dollar sign "$" in the stdout.write method will
+        // print a variable of the same name to the console.
+        if (showResults) stderr.write('$num, ');
```
Cosmetic change, please revert.

rbergen commented on PrimeDart/solution_1/bin/prime_8_bit.dart:
```diff
@@ -126,18 +202,16 @@ class PrimeSieve {
 
     if (showResults) {
-      // Print a new line after the results if the results are shown.
+      // Print a new line after the results if the results are shown.      
```
Whitespace

rbergen commented on PrimeDart/solution_1/bin/prime_8_bit.dart:
```diff
@@ -152,13 +226,13 @@ class PrimeSieve {
 
-    stdout.write(
-        'eagerestwolf&mmcdon20_8bit;$passes;$duration;1;algorithm=base,faithful=yes,bits=8\n');
+    stderr.write(
+        'eagerestwolf&mmcdon20&tarish_8bit;$passes;$duration;1;algorithm=base,faithful=yes,bits=8\n');
```
```suggestion
    stdout.write(
        'eagerestwolf&mmcdon20&tarish_8bit;$passes;$duration;1;algorithm=base,faithful=yes,bits=8\n');
```

rbergen commented on PrimeDart/solution_1/bin/prime_8_bit.dart:
```diff
@@ -152,13 +226,13 @@ class PrimeSieve {
   int countPrimes() {
     var count = (_sieveSize >= 2) ? 1 : 0;
-
+    
```
Whitespace

rbergen commented on PrimeDart/solution_1/bin/prime_8_bit_par.dart:
```diff
@@ -33,6 +33,6 @@ Future<void> main() async {
 
-  stdout.writeln(
-      'eagerestwolf&mmcdon20_8bit_par;$passes;$time;$processors;algorithm=base,faithful=yes,bits=8');
+  stderr.writeln(
+      'eagerestwolf&mmcdon20&tarish_8bit_par;$passes;$time;$processors;algorithm=base,faithful=yes,bits=8');
```
```suggestion
  stdout.writeln(
      'eagerestwolf&mmcdon20&tarish_8bit_par;$passes;$time;$processors;algorithm=base,faithful=yes,bits=8');
```

TarishAhmed commented:
I might be wrong in my approach closing this
